### ISSUE TITLE
Loss Nan

### ISSUE THREAD

**Issue by 0merjavaid** [sagieppel/Fully-convolutional-neural-network-FCN-for-semantic-segmentation-Tensorflow-implementation#1]
I am getting train loss nan at all steps
i'm training it on my own dataset but getting this from start of train.
![image](https://user-images.githubusercontent.com/22716318/34445490-cf19dc00-ecf5-11e7-84de-73954cbfab57.png)

**Comment by sagieppel** [comment]
1) Have you change the number of classes parameter (NUM_CLASSES) to the number of classes in your dataset?
2)Did you check that you dont have nan or inf in your input?

**Comment by 0merjavaid** [comment]
Okay so in my case i have images of retina and I am trying to segment vessels.
so i changed my classes to 1.
further more my input is just 7 images i am trying to overfit for now.
Sample attached.

![image_01l](https://user-images.githubusercontent.com/22716318/34447166-48d46abc-ed03-11e7-95c5-167709b8c5b6.jpg)
![image_01l_1stho](https://user-images.githubusercontent.com/22716318/34447167-490e62d0-ed03-11e7-8d37-75b60918237a.png)

**Comment by 0merjavaid** [comment]
I changed the loss function from sparse softmax to simple softmax
 Loss = tf.reduce_mean((tf.nn.softmax_cross_entropy_with_logits(labels=tf.squeeze(GTLabel, squeeze_dims=[3]),logits=tf.squeeze(Net.Prob,squeeze_dims=[3]),name="loss")))

so it started giving me value in loss.
but loss start to increase instead of decreasing
What is wrong that i am doing?
![image](https://user-images.githubusercontent.com/22716318/34447225-ce6d7a42-ed03-11e7-9139-1ed00e88f6d8.png)

**Comment by sagieppel** [comment]
These values are very very high and Sparse should have worked fine so I suspect you have some problem with the data. This might help https://stackoverflow.com/questions/40050397/tensorflow-nan-loss-reasons
Also training such net with 7 examples is a complete waste of times, you need minimum several hundreds of examples.

**Comment by 0merjavaid** [comment]
well the reason i am using 7 example is to test the net.
to make it overfit on these example so that i train on all data.

**Comment by 0merjavaid** [comment]
since you suspect images
my images are RGB jpegs
and 
ground truth are pngs 
vessels =255(white)
background = 0(black)

**Comment by sagieppel** [comment]
That the problem. If you have two labels your values should be 0,1.
Value of 255 implies 255 classes.

**Comment by 0merjavaid** [comment]
yeah i downloaded your dataset and now i got that
i have changed it to 0 and 1 lets see if it works now

**Comment by 0merjavaid** [comment]
Hats off
your code rocks, it worked Thanks
![image](https://user-images.githubusercontent.com/22716318/34449025-9908845c-ed15-11e7-8286-bb21d5939590.png)



Only one concern left.
dont you think that if you have used tensorflow for augmentation or opencv instead of scipy it would have been much faster?
Since i have felt it myself and seen the results that opencv and tensorflow are faster than scipy
https://www.kaggle.com/zfturbo/test-speed-cv2-vs-scipy-vs-tensorflow

if you want to modify then i can work on it and generate a pull request to speed up operations.

**Comment by sagieppel** [comment]
Cool, glad to hear.
The image loading/augmentation time is usually negligible compared to the backprop time, 
but feel free to add any improvement, that what git for.

**Comment by leishi2018** [comment]
> Hats off
> your code rocks, it worked Thanks
> ![image](https://user-images.githubusercontent.com/22716318/34449025-9908845c-ed15-11e7-8286-bb21d5939590.png)
> 
> Only one concern left.
> dont you think that if you have used tensorflow for augmentation or opencv instead of scipy it would have been much faster?
> Since i have felt it myself and seen the results that opencv and tensorflow are faster than scipy
> https://www.kaggle.com/zfturbo/test-speed-cv2-vs-scipy-vs-tensorflow
> 
> if you want to modify then i can work on it and generate a pull request to speed up operations.

Can you tell me how to solve this problem, my loss is nan and I have change the white 1 and black 0.
and I use the loss function you provide my loss is also very huge and add.

**Comment by sagieppel** [comment]
Have you change the num_classes to 2?
You can send me single image and label map and i will have a look.
